api/internal/git/repospec.go

```diff
@@ -79,99 +80,136 @@ func (x *RepoSpec) Cleaner(fSys filesys.FileSystem) func() error {
 	return func() error { return fSys.RemoveAll(x.Dir.String()) }
 }
 
+const (
+	refQuery         = "?ref="
+	gitSuffix        = ".git"
+	gitRootDelimiter = "_git/"
+	pathSeparator    = string(filepath.Separator)
+)
+
 // NewRepoSpecFromURL parses git-like urls.
 // From strings like git@github.com:someOrg/someRepo.git or
 // https://github.com/someOrg/someRepo?ref=someHash, extract
-// the parts.
+// the different parts of URL, set into a RepoSpec object and return RepoSpec object.
+// It MUST return an error if the input is not a git-like URL, as this is used by some code paths
+// to distinguish between local and remote paths.
+//
+// In particular, NewRepoSpecFromURL separates the URL used to clone the repo from the
+// elements Kustomize uses for other purposes (e.g. query params that turn into args, and
+// the path to the kustomization root within the repo).
 func NewRepoSpecFromURL(n string) (*RepoSpec, error) {
+	repoSpec := &RepoSpec{raw: n, Dir: notCloned, Timeout: defaultTimeout, Submodules: defaultSubmodules}
 	if filepath.IsAbs(n) {
 		return nil, fmt.Errorf("uri looks like abs path: %s", n)
 	}
-	repoSpecVal := parseGitURL(n)
-	if repoSpecVal.RepoPath == "" {
-		return nil, fmt.Errorf("url lacks repoPath: %s", n)
-	}
-	if repoSpecVal.Host == "" {
-		return nil, fmt.Errorf("url lacks host: %s", n)
-	}
-	cleanedPath := filepath.Clean(strings.TrimPrefix(repoSpecVal.KustRootPath, string(filepath.Separator)))
-	if pathElements := strings.Split(cleanedPath, string(filepath.Separator)); len(pathElements) > 0 &&
-		pathElements[0] == filesys.ParentDir {
-		return nil, fmt.Errorf("url path exits repo: %s", n)
-	}
```

Comment:
I initially only moved the `RepoPath` and `Host` validations into their respective parser methods, but then I noticed that this one is effectively just a KustRootPath validation, so I moved it too. Then all I had left here was the `IsAbs` check, so I deleted this and renamed `parseGitURL` to `NewRepoSpecFromURL` (FYI because that's not what it looks like on Github).